### PR TITLE
drivers/serial: Echo input in driver layer

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
Works with https://github.com/apache/nuttx-apps/pull/1608 
## Impact
nsh
## Testing
nsh and telnet

### CODE REVIEW

File: drivers/serial/serial.c [hunk 1]
```diff
@@ -1765,11 +1813,13 @@ int uart_register(FAR const char *path, FAR uart_dev_t *dev)
 
       /* Enable \n -> \r\n translation for the console */
 
-      dev->tc_oflag = OPOST | ONLCR;
+      dev->tc_oflag = OPOST | ONLCR | ECHO;
```
Comment (xiaoxiang781216) [inline drivers/serial/serial.c:1816]: ECHO belong tc_iflag not tc_oflag

**Review by jerpelea — APPROVED** [review]

File: drivers/serial/serial.c [hunk 2]
```diff
@@ -886,6 +888,53 @@ static ssize_t uart_read(FAR struct file *filep,
...
+
+              /* Skipping character count down */
+
+              if (dev->escape > 0)
+                {
+                  dev->escape--;
+                }
+            }
+#ifdef CONFIG_SERIAL_TERMIOS
+          else if (dev->isconsole == false && (dev->tc_iflag & ECHO))
```
Comment (xiaoxiang781216) [inline drivers/serial/serial.c:931]:
```suggestion
          else
#endif
```
and remove line 937

File: drivers/serial/serial.c [hunk 3]
```diff
@@ -886,6 +888,53 @@ static ssize_t uart_read(FAR struct file *filep,
 
           *buffer++ = ch;
           recvd++;
+
+          if (dev->isconsole
+#ifdef CONFIG_SERIAL_TERMIOS
+              && (dev->tc_iflag & ECHO)
```
[the remark this answers is not in the document]
Comment (no1wudi) [reply drivers/serial/serial.c:894]: Why? What if `dev->isconsole` is false and ECHO not set?

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by tmedicci** [comment]
We are still facing issues on ESP32:

nuttx: `14ba7a7dd1e0bbcecf277b1ff230fde6b0a91292`
nuttx-apps: `f1ba1a40ed35d15652d99e8347bf977babe764b3`

```
nsh> 
     nsh> 
          nsh> ls
                 /:
 dev/
 proc/
nsh> ls
       /:
 dev/
 proc/
nsh> 
     nsh> 
          nsh> 
               nsh> ls
                      /:
 dev/
 proc/
nsh>
```

Could you please take a look?

**Comment by lucasssvaz** [comment]
I can also confirm what @tmedicci found.

image

**Comment by no1wudi** [comment]
@tmedicci @lucasssvaz Thanks for your reports，I'll take a look soon

**Comment by tmedicci** [comment]
> @tmedicci @lucasssvaz Thanks for your information，I'll take a look soon.
> 
> Could you share which config used in your report?

Apparently, any configuration throws the same behavior. I used `esp32-devkitc:nsh`.

Maybe we should add the CR when `\n` is detected.

```
diff --git a/drivers/serial/serial.c b/drivers/serial/serial.c
index ddaeb20112..a3f1d41a66 100644
--- a/drivers/serial/serial.c
+++ b/drivers/serial/serial.c
@@ -918,9 +918,11 @@ static ssize_t uart_read(FAR struct file *filep,
 
               if ((!iscntrl(ch & 0xff) || (ch == '\n')) && dev->escape == 0)
                 {
+                  if (ch == '\n')
                     {
-                      uart_putxmitchar(dev, ch, true);
+                      uart_putxmitchar(dev, '\r', true);
                     }
+                  uart_putxmitchar(dev, ch, true);
                 }
 
               /* Skipping character count down */
```

**Comment by no1wudi** [comment]
> > @tmedicci @lucasssvaz Thanks for your information，I'll take a look soon.
> > 
> > Could you share which config used in your report?
> 
> Apparently, any configuration throws the same behavior. I used `esp32-devkitc:nsh`.
> 
> Maybe we should add the CR when `\n` is detected.

Hmm, yes I forget to echo CR before LF in this patch, could you submit this fix?

**Comment by tmedicci** [comment]
> > > @tmedicci @lucasssvaz Thanks for your information，I'll take a look soon.
> > > Could you share which config used in your report?
> > 
> > 
> > Apparently, any configuration throws the same behavior. I used `esp32-devkitc:nsh`.
> > Maybe we should add the CR when `\n` is detected.
> 
> Hmm, yes I forget to echo CR before LF in this patch, could you submit this fix? thanks

Yes! I'm on road, but I'll do it in a couple of hours.

**Comment by slorquet** [comment]
Sorry but this fix breaks my build and has to be reverted
Every character is now echoed twice

using python -m serial --raw --eol cr /dev/ttyUSB0 115200 in linux.

**Comment by slorquet** [comment]
you could be specific and tell me it's in readline.
I am not using nuttx-apps.

**Comment by no1wudi** [comment]
> you could be specific and tell me it's in readline. I am not using nuttx-apps.

If you don't use nuttx-apps, another way is to disable ECHO by termios.

**Comment by slorquet** [comment]
No, it seems that your change in termios to better match posix is a good idea. my apps needed an update.

Please communicate more with the community on the list to warn us that apps must be updated. That would be a 30 seconds email, no need to write a lot. also we will help you to test with corner cases.

It is actually rare to use the official nuttx apps, it is common to use custom local apps that dont automatically follow upstream.

**Comment by no1wudi** [comment]
got it，I'll send a email later.

**Comment by xiaoxiang781216** [comment]
add "breaking change" and "Standard" label.

**Comment by jerpelea** [comment]
I think that all breaking changes should be discussed on the dev mailing list before merging them !

**Comment by TimJTi** [comment]
I know there is lots of discussion...just rebased a branch to master to submit a PR and now have double characters in my console that has behaved for years.

REVERT UNTIL SORTED!!

**Comment by no1wudi** [comment]
> I know there is lots of discussion...just rebased a branch to master to submit a PR and now have double characters in my console that has behaved for years.
> 
> REVERT UNTIL SORTED!!


Please rebase nuttx-apps to master also, I know it's a mistake that merge it before fully discussed.

But it will break many other peoples again if revert it now. The discussion is on going in mail list, let's wait for the conclusion to revert it or not.

**Comment by TimJTi** [comment]
This is nothing to do with apps. The basic console is broken with echos I can't suppress. I'm assuming it's this that has broken it but have not spent time to bisect

**Comment by no1wudi** [comment]
@TimJTi Please refer to https://github.com/apache/nuttx-apps/pull/1608, the echo is done by application layer before, but it echoed by serial driver with this patch, so the nuttx-apps should update together to let nsh work correctly.
